neqo-transport/tests/sim/connection.rs

```diff
@@ -51,28 +64,44 @@ impl ConnectionNode {
 }
 
 impl Node for ConnectionNode {
-    fn process(&mut self, d: Option<Datagram>, now: Instant) -> Output {
-        let res = self.c.process(d, now);
-        for e in self.c.events() {
-            qdebug!([self.c], "received event {:?}", e);
-
-            // Perform authentication automatically.
-            if matches!(e, ConnectionEvent::AuthenticationNeeded) {
-                self.c.authenticated(AuthenticationStatus::Ok, now);
-            }
+    fn init(&mut self, now: Instant) {
+        for g in &mut self.goals {
+            g.init(&mut self.c, now);
+        }
+    }
 
-            // Waiting on drain_filter...
-            // let _ = self.goals.drain_filter(|g| g.handle_event(&mut self.c, &e)).count();
-            let mut i = 0;
-            while i < self.goals.len() {
-                if self.goals[i].handle_event(&mut self.c, &e, now) {
-                    self.goals.remove(i);
-                } else {
-                    i += 1;
+    fn process(&mut self, mut d: Option<Datagram>, now: Instant) -> Output {
+        loop {
+            let res = self.c.process(d.take(), now);
+            let mut active = false;
+            while let Some(e) = self.c.next_event() {
+                qdebug!([self.c], "received event {:?}", e);
+
+                // Perform authentication automatically.
+                if matches!(e, ConnectionEvent::AuthenticationNeeded) {
+                    self.c.authenticated(AuthenticationStatus::Ok, now);
+                }
+
+                // Waiting on drain_filter...
+                // let _ = self.goals.drain_filter(|g| g.handle_event(&mut self.c, &e)).count();
+                let mut i = 0;
+                while i < self.goals.len() {
+                    let status = self.goals[i].handle_event(&mut self.c, &e, now);
+                    if status == GoalStatus::Done {
+                        self.goals.remove(i);
+                    } else {
+                        active |= status == GoalStatus::Active;
+                        i += 1;
+                    }
                 }
             }
+            // Exit at this point if the connection produced a datagram.
+            // OR if one of the goals acted.
+            if matches!(res, Output::Datagram(_)) || !active {
```

Comment:
No, I think that the comment explains this:

So if we get a datagram, that is easy.  We have to stop.

But we want to make sure that if any of the goals report activity and we don't have any datagram, that might have changed the connection state so that we will get a datagram (or at least we might get a different timeout).

I will amend the comment (and correct a logic error).